persist/fs/index_read.go

```diff
@@ -0,0 +1,383 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"github.com/m3db/m3db/digest"
+	"github.com/m3db/m3db/generated/proto/index"
+	"github.com/m3db/m3db/persist"
+	"github.com/m3db/m3db/x/mmap"
+
+	xlog "github.com/m3db/m3x/log"
+)
+
+type indexReader struct {
+	opts           Options
+	filePathPrefix string
+	hugePagesOpts  mmap.HugeTLBOptions
+	logger         xlog.Logger
+
+	start        time.Time
+	namespaceDir string
+
+	currIdx                int
+	info                   index.IndexInfo
+	expectedDigest         index.IndexDigests
+	expectedDigestOfDigest uint32
+	readDigests            indexReaderReadDigests
+}
+
+type indexReaderReadDigests struct {
+	infoFileDigest    uint32
+	digestsFileDigest uint32
+	segments          []indexReaderReadSegmentDigests
+}
+
+type indexReaderReadSegmentDigests struct {
+	segmentType IndexSegmentType
+	files       []indexReaderReadSegmentFileDigest
+}
+
+type indexReaderReadSegmentFileDigest struct {
+	segmentFileType IndexSegmentFileType
+	digest          uint32
+}
+
+// NewIndexReader returns a new index reader with options.
+func NewIndexReader(opts Options) (IndexFileSetReader, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+	r := new(indexReader)
+	r.reset(opts)
+	return r, nil
+}
+
+func (r *indexReader) reset(opts Options) {
+	*r = indexReader{}
+	r.opts = opts
+	r.filePathPrefix = opts.FilePathPrefix()
+	r.hugePagesOpts = mmap.HugeTLBOptions{
+		Enabled:   opts.MmapEnableHugeTLB(),
+		Threshold: opts.MmapHugeTLBThreshold(),
+	}
+	r.logger = opts.InstrumentOptions().Logger()
+}
+
+func (r *indexReader) Open(opts IndexReaderOpenOptions) error {
+	r.reset(r.opts)
```

Comment:
is this required?

Reply:
If you want to be able to reuse the reader it is yes, we do tend to do that a fair bit.

Reply:
Fair enough